lib/ansible/modules/cloud/amazon/s3.py

```diff
@@ -272,131 +240,140 @@
     if module.check_mode:
         module.exit_json(msg="PUT operation skipped - running in check mode", changed=True)
     if location is None:
-        location = Location.DEFAULT
+        location = 'us-east-1'
+    configuration = {}
+    if location != ('us-east-1' or None):
+        configuration['LocationConstraint'] = location
     try:
-        bucket = s3.create_bucket(bucket, location=location)
+        if len(configuration) > 0:
+            s3.create_bucket(Bucket=bucket, CreateBucketConfiguration=configuration)
+        else:
+            s3.create_bucket(Bucket=bucket)
         for acl in module.params.get('permission'):
-            bucket.set_acl(acl)
-    except s3.provider.storage_response_error as e:
-        module.fail_json(msg="Failed while creating bucket or setting acl (check that you have CreateBucket and PutBucketAcl permission) %s: %s" % (bucket, e),
-                exception=traceback.format_exc())
+            s3.put_bucket_acl(ACL=acl, Bucket=bucket)
+    except botocore.exceptions.ClientError as e:
+        module.fail_json(msg="Failed while creating bucket or setting acl (check that you have CreateBucket and PutBucketAcl permission).",
+                         exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
+
     if bucket:
         return True
 
-def get_bucket(module, s3, bucket):
-    try:
-        return s3.lookup(bucket)
-    except s3.provider.storage_response_error as e:
-        module.fail_json(msg="Failed while getting bucket %s: %s" % (bucket, e),
-                exception=traceback.format_exc())
-
-def list_keys(module, bucket_object, prefix, marker, max_keys):
-    all_keys = bucket_object.get_all_keys(prefix=prefix, marker=marker, max_keys=max_keys)
-
-    keys = [x.key for x in all_keys]
 
+def list_keys(module, s3, bucket, prefix, marker, max_keys):
+    all_keys = s3.list_objects(Bucket=bucket, Prefix=prefix, Marker=marker, MaxKeys=max_keys)
+    keys = [x['Key'] for x in all_keys.get('Contents', [])]
     module.exit_json(msg="LIST operation complete", s3_keys=keys)
 
+
 def delete_bucket(module, s3, bucket):
     if module.check_mode:
         module.exit_json(msg="DELETE operation skipped - running in check mode", changed=True)
     try:
-        bucket = s3.lookup(bucket, validate=False)
-        bucket_contents = bucket.list()
-        bucket.delete_keys([key.name for key in bucket_contents])
-    except s3.provider.storage_response_error as e:
-        if e.status == 404:
-            # bucket doesn't appear to exist
+        exists = bucket_check(module, s3, bucket)
+        if exists is False:
             return False
-        elif e.status == 403:
-            # bucket appears to exist but user doesn't have list bucket permission; may still be able to delete bucket
-            pass
-        else:
-            module.fail_json(msg=str(e), exception=traceback.format_exc())
-    try:
-        bucket.delete()
+        objects = s3.list_objects(Bucket=bucket)
```

Comment:
I have a concern on pagination here, `list_objects` is going to default to 100 or 500 items, which is fine most of the time (it was like this prior) but would make a nice enhancement.